File: app/src/main/java/com/airadar/app/data/FlightStore.kt
package com.airadar.app.data

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.time.Instant
import java.time.LocalDate
import java.time.temporal.ChronoUnit

/**
 * In-memory source of truth shared by the trip list and the importers. Signed in,
 * it mirrors the account on the server; signed out, it holds only this session's
 * additions and starts empty.
 */
object FlightStore {

    /** How long a deleted trip waits in the recycle bin before it is gone for good. */
    const val RETENTION_DAYS = 30L

    // Everything, deleted trips included; the two LiveData below are its two halves.
    private var all: List<Flight> = emptyList()

    private val _flights = MutableLiveData<List<Flight>>(emptyList())
    val flights: LiveData<List<Flight>> = _flights

    private val _deleted = MutableLiveData<List<Flight>>(emptyList())
    val deleted: LiveData<List<Flight>> = _deleted

    /** The last plan refusal, for the UI to explain and offer an upgrade. */
    private val _limitHit = MutableLiveData<LimitReached?>(null)
    val limitHit: LiveData<LimitReached?> = _limitHit

    fun clearLimitHit() {
        _limitHit.value = null
    }

    /** Every leg actually flown (or in the air right now) -- what the
     * milestone tiers count against. The same filter My's own map uses for
     * its route network, kept in one place so the two never drift apart. */
    val completedFlights: List<Flight>
        get() = (_flights.value ?: emptyList()).filter {
            (it.phase == FlightPhase.PAST || it.phase == FlightPhase.IN_PROGRESS) && !it.isPending
        }

    /** Where the tier ladder puts this traveller right now. */
    val tierStanding: TierStanding get() = TierStanding.compute(completedFlights)

    /** Rainbow's sequence number, once claimed -- "the Nth traveller here".
     * Never set locally: only the backend hands one out, and only once
     * standing genuinely reads Rainbow. */
    private val _rainbowRank = MutableLiveData<Int?>(null)
    val rainbowRank: LiveData<Int?> = _rainbowRank

    /** Safe to call any time standing is checked: an existing claim just
     * comes back unchanged, so there's no harm calling this opportunistically. */
    fun refreshRainbowRank() {
        if (tierStanding.tier != MilestoneTier.RAINBOW || !synced) return
        scope.launch {
            val rank = runCatching { BackendClient.claimRainbow() }.getOrNull()
            withContext(Dispatchers.Main) { _rainbowRank.value = rank }
        }
    }

    // Server writes go out from here, off the main thread, one after another.
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val synced: Boolean get() = AuthStore.isSignedIn


    /** Pull-to-refresh: the account's trips when signed in, the demo set otherwise. */
    suspend fun refresh() {
        if (synced) syncFromServer() else reseed()
    }

    /**
     * Replaces everything with what the account holds. Called at sign-in and on
     * refresh; the demo trips shown while signed out are not carried over.
     */
    suspend fun syncFromServer() {
        val outgoing = runCatching { BackendClient.outgoingShares() }.getOrDefault(emptyMap())
        val live = BackendClient.listTrips().map { it.copy(shares = outgoing[it.id].orEmpty()) }
        val binned = BackendClient.listTrips(deleted = true)
        val incoming = runCatching { BackendClient.incomingShares() }.getOrDefault(emptyList())
        // A trip taken "together" is mine now (the server copied it); the friend's
        // name rides on my copy instead of a second card.
        val together = incoming.filter { it.sharedBy?.status == ShareStatus.TOGETHER }
        val mine = live.map { own ->
            together.firstOrNull { it.flightNumber == own.flightNumber && it.departureTime == own.departureTime }
                ?.let { own.copy(sharedBy = it.sharedBy) } ?: own
        }
        val rest = incoming.filterNot { inc ->
            inc.sharedBy?.status == ShareStatus.TOGETHER &&
                    live.any { it.flightNumber == inc.flightNumber && it.departureTime == inc.departureTime }
        }
        withContext(Dispatchers.Main) { publish(mine + binned + rest) }
    }

    /** Accept, reject or take together a friend's trip; the list is then refreshed. */
    suspend fun respondToShare(flight: Flight, action: String) {
        val share = flight.sharedBy ?: return
        BackendClient.respondToShare(share.id, action)
        syncFromServer()
    }

    /** Shares one of my trips with a friend and shows the new block straight away. */
    suspend fun shareWith(flight: Flight, person: Person) {
        val (share, _) = BackendClient.shareTrip(flight.id, person.id)
        if (share != null) {
            withContext(Dispatchers.Main) {
                publish(all.map { f ->
                    if (f.id == flight.id) f.copy(shares = (f.shares.filterNot { it.person.id == person.id } + share)) else f
                })
            }
        }
    }

    /** Back to the signed-out demo list. */
    fun onSignedOut() {
        all = emptyList()
        reseed()
    }

    private fun push(block: suspend () -> Unit) {
        if (!synced) return
        scope.launch { runCatching { block() } }
    }

    /**
     * Signed out, the list is whatever the traveller added on this phone — the
     * app starts empty. A refresh re-publishes it (recomputing typical durations
     * and dropping expired bin entries) and nothing more.
     */
    fun reseed() {
        publish(all)
    }

    /** Adds a trip; returns false (and records why) when the plan does not allow it. */
    fun add(flight: Flight): Boolean {
        val same = all.firstOrNull { it.flightNumber == flight.flightNumber && it.departureTime == flight.departureTime }
        if (same == null) {
            try {
                Entitlements.checkAdd(flight, all)
            } catch (e: LimitReached) {
                _limitHit.value = e
                return false
            }
        }
        when {
            same == null -> {
                publish(all + flight)
                push {
                    try {
                        BackendClient.putTrip(flight)
                    } catch (e: BackendClient.BackendException) {
                        // The server's count is the truth; take the trip back out.
                        if (e.code == 402) withContext(Dispatchers.Main) {
                            publish(all.filterNot { it.id == flight.id })
                            _limitHit.value = LimitReached(e.message ?: "Plan limit reached.", Entitlements.membership.tier)
                        } else throw e
                    }
                }
            }
            // Adding a trip that sits in the bin brings it back rather than duplicating it.
            same.deletedAt != null -> restore(same.id)
        }
        return true
    }

    fun confirm(flight: Flight) {
        val confirmed = flight.copy(isPending = false)
        publish(all.map { if (it.id == flight.id) confirmed else it })
        push { BackendClient.putTrip(confirmed) }
    }

    fun replace(old: Flight, replacement: Flight) {
        val confirmed = replacement.copy(isPending = false)
        publish(all.filterNot { it.id == old.id } + confirmed)
        push {
            if (old.id != confirmed.id) BackendClient.deleteTrip(old.id)
            BackendClient.putTrip(confirmed)
        }
    }

    /** Moves the trip to the recycle bin; [restore] undoes it within [RETENTION_DAYS]. */
    fun delete(flightId: String) {
        val target = all.firstOrNull { it.id == flightId }
        // A friend's trip has no bin: swiping it away is declining the share.
        // (A trip taken together is my own copy and goes to the bin like any other.)
        target?.sharedBy?.takeIf { flightId.startsWith("shared:") }?.let { share ->
            publish(all.filterNot { it.id == flightId })
            push { BackendClient.respondToShare(share.id, "reject") }
            return
        }
        publish(all.map { if (it.id == flightId) it.copy(deletedAt = Instant.now()) else it })
        push { BackendClient.deleteTrip(flightId) }
    }

    fun restore(flightId: String) {
        publish(all.map { if (it.id == flightId) it.copy(deletedAt = null) else it })
        push { BackendClient.restoreTrip(flightId) }
    }

    fun setTrack(flightId: String, points: List<TrackPoint>, flownOn: LocalDate) {
        publish(all.map { if (it.id == flightId) it.copy(track = points, trackFlownOn = flownOn) else it })
        all.firstOrNull { it.id == flightId }?.let { updated -> push { BackendClient.putTrip(updated) } }
    }

    private fun publish(list: List<Flight>) {
        val expiry = Instant.now().minus(RETENTION_DAYS, ChronoUnit.DAYS)
        val live = list.filter { it.deletedAt == null }
        // Typical durations are averaged over live trips only.
        all = list
            .filterNot { it.deletedAt?.isBefore(expiry) == true }
            .map { it.copy(typicalDurationMinutes = typicalDuration(it, live)) }
        _flights.value = all.filter { it.deletedAt == null }.sortedBy { it.departureInstant ?: Instant.MAX }
        _deleted.value = all.filter { it.deletedAt != null }.sortedByDescending { it.deletedAt }
    }

    /**
     * Average block time over the last week of the same flight number, falling back
     * to the scheduled duration when there is nothing recent to average.
     */
    private fun typicalDuration(flight: Flight, pool: List<Flight>): Int {
        val cutoff = Instant.now().minus(7, ChronoUnit.DAYS)
        val recent = pool.filter {
            it.flightNumber == flight.flightNumber &&
                    it.phase == FlightPhase.PAST &&
                    (it.departureInstant ?: Instant.MIN).isAfter(cutoff)
        }
        return if (recent.isEmpty()) flight.durationMinutes
        else recent.sumOf { it.durationMinutes + it.delayMinutes } / recent.size
    }
}

File: ios/Airadar/Data/FlightStore.swift
import Foundation
import Combine

/// In-memory source of truth shared by every screen. Signed in, it mirrors the
/// account on the server; signed out, it holds only this session's additions and
/// starts empty.
@MainActor
final class FlightStore: ObservableObject {
    static let shared = FlightStore()

    /// How long a deleted trip waits in the recycle bin before it is gone for good.
    static let retentionDays = 30

    @Published private(set) var flights: [Flight] = [] {
        // The Home Screen icon follows the tier this recomputes to — a
        // no-op call whenever it hasn't actually changed (AppIconManager
        // checks before ever touching UIApplication).
        didSet { AppIconManager.apply(tierStanding.tier) }
    }
    @Published private(set) var deleted: [Flight] = []
    /// The last plan refusal, for the UI to explain and offer an upgrade.
    @Published var limitHit: LimitReached?

    // Everything, deleted trips included; the two lists above are its two halves.
    private var all: [Flight] = []

    private var synced: Bool { AuthStore.shared.isSignedIn }

    /// A flight within a day and a half of now: worth polling for status.
    var hasFlightNearNow: Bool {
        let now = Date()
        return flights.contains { f in
            guard let dep = f.departureInstant else { return false }
            return abs(dep.timeIntervalSince(now)) < 36 * 3600
        }
    }

    /// Every leg actually flown (or in the air right now) — what the
    /// milestone tiers count against. The same filter My's own map uses for
    /// its route network, kept in one place so the two never drift apart.
    var completedFlights: [Flight] {
        flights.filter { ($0.phase == .past || $0.phase == .inProgress) && !$0.isPending }
    }
    var completedFlightCount: Int { completedFlights.count }

    /// Where the tier ladder puts this traveller right now.
    var tierStanding: TierStanding { .compute(for: completedFlights) }

    /// Porcelain's sequence number, once claimed — "the Nth traveller here".
    /// Never set locally: only the backend hands one out, and only once
    /// standing genuinely reads Porcelain.
    @Published private(set) var porcelainRank: Int?

    /// Safe to call any time standing is checked: an existing claim just
    /// comes back unchanged, so there's no harm calling this opportunistically.
    func refreshPorcelainRank() async {
        guard tierStanding.tier == .porcelain, AuthStore.shared.isSignedIn else { return }
        porcelainRank = try? await BackendClient.claimPorcelain()
    }

    // The last published list, on disk, so a relaunch shows the trips at once and
    // the server sync only refines them.
    nonisolated private static let cacheURL: URL = {
        let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.appendingPathComponent("trips.json")
    }()

    private init() {
        if let data = try? Data(contentsOf: Self.cacheURL),
           let cached = try? BackendClient.decoder.decode([Flight].self, from: data) {
            publish(cached)
        }
    }

    private func persist() {
        let snapshot = all
        let url = Self.cacheURL
        Task.detached(priority: .utility) {
            if let data = try? BackendClient.encoder.encode(snapshot) { try? data.write(to: url, options: .atomic) }
        }
    }

    /// Pull-to-refresh: the account's trips when signed in, a re-publish otherwise.
    func refresh() async {
        if synced { try? await syncFromServer() } else { publish(all) }
    }

    /// Replaces everything with what the account holds. Called at sign-in and on refresh.
    func syncFromServer() async throws {
        let outgoing = (try? await BackendClient.outgoingShares()) ?? [:]
        var live = try await BackendClient.listTrips()
        for i in live.indices { live[i].shares = outgoing[live[i].id] ?? [] }
        let binned = try await BackendClient.listTrips(deleted: true)
        let incoming = (try? await BackendClient.incomingShares()) ?? []

        // A trip taken "together" is mine now (the server copied it); the friend's
        // name rides on my copy instead of a second card.
        let together = incoming.filter { $0.sharedBy?.status == .together }
        var mine = live
        for i in mine.indices {
            if let t = together.first(where: { $0.flightNumber == mine[i].flightNumber && $0.departureTime == mine[i].departureTime }) {
                mine[i].sharedBy = t.sharedBy
            }
        }
        let rest = incoming.filter { inc in
            !(inc.sharedBy?.status == .together &&
              live.contains { $0.flightNumber == inc.flightNumber && $0.departureTime == inc.departureTime })
        }
        // What moved since the last look — a delay, a gate, a status — is announced.
        let before = Dictionary(uniqueKeysWithValues: all.map { ($0.id, $0) })
        for f in mine { if let old = before[f.id] { FlightReminders.announceChange(from: old, to: f) } }
        publish(mine + binned + rest)
        for f in mine where f.phase == .upcoming { SiriSuggestions.donate(f) }
    }

    /// Back to the signed-out list: empty.
    func onSignedOut() {
        all = []
        publish(all)
        try? FileManager.default.removeItem(at: Self.cacheURL)
    }

    private func push(_ block: @escaping @Sendable () async throws -> Void) {
        guard synced else { return }
        Task { try? await block() }
    }

    /// Adds a trip; returns false (and records why) when the plan does not allow it.
    @discardableResult
    func add(_ flight: Flight) -> Bool {
        let same = all.first { $0.flightNumber == flight.flightNumber && $0.departureTime == flight.departureTime }
        if same == nil {
            do { try Entitlements.checkAdd(flight, existing: all) } catch {
                limitHit = error
                return false
            }
        }
        if let same {
            // Adding a trip that sits in the bin brings it back rather than duplicating it.
            if same.deletedAt != nil { restore(same.id) }
            return true
        }
        publish(all + [flight])
        SiriSuggestions.donate(flight)
        if synced {
            Task { [weak self] in
                do {
                    let stored = try await BackendClient.putTrip(flight)
                    // A fresh fed flight (see Flight.feedStatus) gets its real
                    // review outcome back in this same response — reflect it
                    // at once, rather than leaving the card reading "pending"
                    // until whatever's next syncs from the server.
                    guard let self, stored.feedStatus != flight.feedStatus else { return }
                    self.publish(self.all.map { $0.id == stored.id ? stored : $0 })
                } catch let e as BackendClient.BackendError where e.code == 402 {
                    // The server's count is the truth; take the trip back out.
                    guard let self else { return }
                    self.publish(self.all.filter { $0.id != flight.id })
                    self.limitHit = LimitReached(reason: e.message, tier: Entitlements.membership.tier)
                } catch {}
            }
        }
        return true
    }

    func confirm(_ flight: Flight) {
        // A `let` copy: a @Sendable closure may not capture a `var`.
        let confirmed = flight.confirmed()
        publish(all.map { $0.id == flight.id ? confirmed : $0 })
        SiriSuggestions.donate(confirmed)
        push { try await BackendClient.putTrip(confirmed) }
    }

    func replace(_ old: Flight, with replacement: Flight) {
        let confirmed = replacement.confirmed()
        let oldId = old.id
        publish(all.filter { $0.id != oldId } + [confirmed])
        push {
            if oldId != confirmed.id { try await BackendClient.deleteTrip(oldId) }
            try await BackendClient.putTrip(confirmed)
        }
    }

    /// Moves the trip to the recycle bin; a friend's shared trip is declined instead.
    func delete(_ id: String) {
        if id.hasPrefix("shared:"), let share = all.first(where: { $0.id == id })?.sharedBy {
            publish(all.filter { $0.id != id })
            push { try await BackendClient.respondToShare(share.id, action: "reject") }
            return
        }
        publish(all.map { var f = $0; if f.id == id { f.deletedAt = Date() }; return f })
        SiriSuggestions.forget(id)
        push { try await BackendClient.deleteTrip(id) }
    }

    /// Every current trip, one at a time — the same as swiping each away, all
    /// at once: a real trip goes to the Recycle Bin, a friend's shared one is
    /// just declined, exactly as a single delete already does either way.
    func deleteAll() {
        for f in flights { delete(f.id) }
    }

    func restore(_ id: String) {
        publish(all.map { var f = $0; if f.id == id { f.deletedAt = nil }; return f })
        push { try await BackendClient.restoreTrip(id) }
    }

    func setTrack(_ id: String, points: [TrackPoint], flownOn: String, icao24: String? = nil) {
        publish(all.map { var f = $0; if f.id == id { f.track = points; f.trackFlownOn = flownOn }; return f })
        if let updated = all.first(where: { $0.id == id }) { push { try await BackendClient.putTrip(updated) } }
        // The airframe's own hex, from the track just fetched — worth a free,
        // one-off aircraft-type lookup if the trip doesn't already have one.
        if let icao24, all.first(where: { $0.id == id })?.aircraft == nil {
            Task { await self.fillAircraft(id, icao24: icao24) }
        }
    }

    /// adsbdb, keyed by Mode-S hex — free, and separate from AeroDataBox's
    /// quota entirely. Only ever fills a gap, never overwrites a real value.
    private func fillAircraft(_ id: String, icao24: String) async {
        guard let found = try? await AdsbdbClient.shared.aircraft(modeS: icao24),
              let type = found.type ?? found.icaoType else { return }
        publish(all.map { f in var g = f; if g.id == id, g.aircraft == nil { g.aircraft = type }; return g })
        if let updated = all.first(where: { $0.id == id }) { push { try await BackendClient.putTrip(updated) } }
    }

    /// A callsign adsbdb resolved that the trip didn't already have — the
    /// static bundled table only knows a few dozen airlines; this reaches any
    /// of them. Filled in and synced like a track is, so it is asked for once.
    func applyCallsign(_ id: String, callsign: String) {
        publish(all.map { f in var g = f; if g.id == id, g.callsign == nil { g.callsign = callsign }; return g })
        if let updated = all.first(where: { $0.id == id }) { push { try await BackendClient.putTrip(updated) } }
    }

    /// Accept, reject or take together a friend's trip; the list is then refreshed.
    func respondToShare(_ flight: Flight, action: String) async throws {
        guard let share = flight.sharedBy else { return }
        try await BackendClient.respondToShare(share.id, action: action)
        try await syncFromServer()
    }

    /// Shares one of my trips with a friend and shows the new block straight away.
    func share(_ flight: Flight, with person: Person) async throws {
        let (share, _) = try await BackendClient.shareTrip(flight.id, to: person.id)
        guard let share else { return }
        publish(all.map { f in
            guard f.id == flight.id else { return f }
            var g = f
            g.shares = g.shares.filter { $0.person.id != person.id } + [share]
            return g
        })
    }

    private func publish(_ list: [Flight]) {
        let expiry = Date().addingTimeInterval(-Double(Self.retentionDays) * 86400)
        let live = list.filter { $0.deletedAt == nil }
        all = list
            .filter { !($0.deletedAt.map { $0 < expiry } ?? false) }
            .map { var f = $0; f.typicalDurationMinutes = typicalDuration(f, pool: live); return f }
        flights = all.filter { $0.deletedAt == nil }.sorted { ($0.departureInstant ?? .distantFuture) < ($1.departureInstant ?? .distantFuture) }
        deleted = all.filter { $0.deletedAt != nil }.sorted { ($0.deletedAt ?? .distantPast) > ($1.deletedAt ?? .distantPast) }
        persist()
        LiveActivities.sync(flights)
        WatchSync.shared.sync(flights)
    }

    /// Average block time over the last week of the same number, else the scheduled duration.
    private func typicalDuration(_ flight: Flight, pool: [Flight]) -> Int {
        let cutoff = Date().addingTimeInterval(-7 * 86400)
        let recent = pool.filter {
            $0.flightNumber == flight.flightNumber && $0.phase == .past && ($0.departureInstant ?? .distantPast) > cutoff
        }
        if recent.isEmpty { return flight.durationMinutes }
        return recent.reduce(0) { $0 + $1.durationMinutes + $1.delayMinutes } / recent.count
    }
}

private extension Flight {
    func confirmed() -> Flight { var f = self; f.isPending = false; return f }
}
